connectivity/netsocket/source/InternetDatagramSocket.cpp

```diff
@@ -27,6 +27,53 @@ nsapi_error_t InternetDatagramSocket::connect(const SocketAddress &address)
     return NSAPI_ERROR_OK;
 }
 
+nsapi_size_or_error_t InternetDatagramSocket::sendmsg(const SocketAddress &address, const void *data, nsapi_size_t size, nsapi_msghdr_t *control, nsapi_size_t control_size)
+{
+    _lock.lock();
+    nsapi_size_or_error_t ret;
+
+    _writers++;
+    if (_socket) {
+        _socket_stats.stats_update_socket_state(this, SOCK_OPEN);
+        _socket_stats.stats_update_peer(this, address);
+    }
+    while (true) {
+        if (!_socket) {
+            ret = NSAPI_ERROR_NO_SOCKET;
+            break;
+        }
+
+        core_util_atomic_flag_clear(&_pending);
+        nsapi_size_or_error_t sent = _stack->socket_sendmsg(_socket, address, data, size, control, control_size);
+        if ((0 == _timeout) || (NSAPI_ERROR_WOULD_BLOCK != sent)) {
+            _socket_stats.stats_update_sent_bytes(this, sent);
+            ret = sent;
+            break;
+        } else {
+            uint32_t flag;
+
+            // Release lock before blocking so other threads
+            // accessing this object aren't blocked
+            _lock.unlock();
+            flag = _event_flag.wait_any(WRITE_FLAG, _timeout);
+            _lock.lock();
+
+            if (flag & osFlagsError) {
+                // Timeout break
+                ret = NSAPI_ERROR_WOULD_BLOCK;
+                break;
+            }
+        }
+    }
+
+    _writers--;
+    if (!_socket || !_writers) {
+        _event_flag.set(FINISHED_FLAG);
+    }
+    _lock.unlock();
+    return ret;
+}
+
 nsapi_size_or_error_t InternetDatagramSocket::sendto(const SocketAddress &address, const void *data, nsapi_size_t size)
```

Comment:
`sendto` can call into `sendmsg` to reduce code size.

Reply:
May not reduce code size if not using the metadata stuff.  It might be hard to avoid it anyway, but it's possible that `sendto` could use a direct path to something that avoids some metadata handling.

Reply:
Done